enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -180,4 +188,71 @@ impl<'a> Handler<'a> {
             ret
         }
     }
+
+    /// Do a readv() syscall
+    pub fn readv(&mut self) -> u64 {
+        let fd = self.aex.gpr.rdi;
+        let iovec = self.aex.gpr.rsi as *mut u8;
+        let iovcnt = self.aex.gpr.rdx as usize;
+
+        // Add up total size of buffers.
+        let iovec = unsafe { from_raw_parts_mut(iovec as *mut Iovec, iovcnt) }; // ???
+        let mut bufsize: usize = 0;
+        for i in 0..iovcnt {
+            bufsize += iovec[i].size;
+        }
+
+        // Allocate some unencrypted memory.
+        let size = bufsize + size_of::<Iovec>() * iovcnt;
+        let map = match self.ualloc(size as u64) {
+            Err(errno) => return errno.into_syscall(),
+            Ok(map) => map,
+        };
+
+        // Cast beginning of allocated memory as iovec slice of length iovcnt.
+        let map = unsafe { from_raw_parts_mut(map as *mut Iovec, iovcnt) };
+
+        // Set pointers in unencrypted iovec slice to use the rest of the allocated memory.
+        let mut chunk = map.as_ptr() as usize + size_of::<Iovec>() * iovcnt; // ??? May be unsafe or illegal as per from_raw_parts_mut rules
+        for i in 0..iovcnt {
+            if chunk > map.as_ptr() as usize + size {
+                // ???
+                // TODO: What error to return?
+                self.attacked();
+            }
+            map[i].base = chunk as *mut u8;
+            chunk = chunk + iovec[i].size;
+        }
+
+        unsafe {
+            // Do the syscall; replace encrypted memory with unencrypted memory.
+            let ret = self.syscall(
+                SysCall::READV,
+                fd,
+                map.as_ptr() as _,
+                iovcnt as u64,
+                0,
+                0,
+                0,
+            ); // ???
+            self.ufree(map.as_ptr() as *mut u8, size as u64); // ???
```

Comment:
Don't free this until after you are done copying it...

Reply:
You mean after `copy_nonoverlapping`, right? I was confused about this because in `read()` the `ufree` is called immediately after the syscall, which I didn't understand.

Reply:
@lkatalin You probably don't understand it because it probably doesn't work. I was a bad boy and didn't test `read()` obviously... :)

Reply:
Fair enough!